app/uk/gov/hmrc/soletraderidentificationfrontend/models/SoleTraderDetails.scala

```diff
@@ -52,38 +53,85 @@ object SoleTraderDetails {
-    ) (SoleTraderDetails.apply _)
-
-  val writes: OWrites[SoleTraderDetails] = (
-    (JsPath \ FullNameKey).write[FullName] and
-      (JsPath \ DateOfBirthKey).write[LocalDate] and
-      (JsPath \ NinoKey).writeNullable[String] and
-      (JsPath \ AddressKey).writeNullable[Address] and
-      (JsPath \ SaPostcodeKey).writeNullable[String] and
-      (JsPath \ SautrKey).writeNullable[String] and
-      (JsPath \ IdentifiersMatchKey).write[SoleTraderDetailsMatchResult] and
-      (JsPath \ BusinessVerificationKey).writeNullable[BusinessVerificationStatus] and
-      (JsPath \ RegistrationKey).writeNullable[RegistrationStatus] and
-      (JsPath \ TrnKey).writeNullable[String] and
-      (JsPath \ OverseasIdentifiersKey).writeNullable[Overseas] and
-      (JsPath \ ReputationKey).writeNullable[JsObject]
-    ) (unlift(SoleTraderDetails.unapply))
-
-  implicit val format: OFormat[SoleTraderDetails] = OFormat(reads, writes)
+
+  implicit val format: OFormat[SoleTraderDetails] = new OFormat[SoleTraderDetails] {
+    override def reads(json: JsValue): JsResult[SoleTraderDetails] =
+      for {
+        fullName <- (json \ FullNameKey).validate[FullName]
+        dateOfBirthKey <- (json \ DateOfBirthKey).validate[LocalDate]
+        optNino <- (json \ NinoKey).validateOpt[String]
+        optAddress <- (json \ AddressKey).validateOpt[Address]
+        optSaPostcode <- (json \ SaPostcodeKey).validateOpt[String]
+        optSaUtr <- (json \ SautrKey).validateOpt[String]
+        identifiersMatch <- (json \ IdentifiersMatchKey).validate[SoleTraderDetailsMatchResult]
+        businessVerification <- (json \ BusinessVerificationKey).validateOpt[BusinessVerificationStatus]
+        registrationStatus <- (json \ RegistrationKey).validateOpt[RegistrationStatus]
+        optTrnKey <- (json \ TrnKey).validateOpt[String]
+        optOverseas <- (json \ OverseasIdentifiersKey).validateOpt[Overseas]
+        optOverseasTaxId <- (json \ OverseasTaxIdentifierKey).validateOpt[String]
+        optOverseasTaxIdCountry <- (json \ OverseasCountryKey).validateOpt[String]
+        reputationKey <- (json \ ReputationKey).validateOpt[JsObject]
+      } yield {
+        val (overseasTaxId, overseasTaxIdCountry) = determineOverseasTaxIdentifierDetails(
+          optOverseasTaxId,
+          optOverseasTaxIdCountry,
+          optOverseas
+        )
+        SoleTraderDetails(fullName, dateOfBirthKey, optNino, optAddress, optSaPostcode,
+          optSaUtr, identifiersMatch, businessVerification, registrationStatus, optTrnKey,
+          overseasTaxId, overseasTaxIdCountry, reputationKey)
+      }
+
+    override def writes(soleTraderDetails: SoleTraderDetails): JsObject =
+      Json.obj(
+        FullNameKey -> soleTraderDetails.fullName,
+        DateOfBirthKey -> soleTraderDetails.dateOfBirth,
+        IdentifiersMatchKey -> soleTraderDetails.identifiersMatch.toString,
+      ) ++ {
+        soleTraderDetails.registrationStatus match {
```

Comment:
You need to add the two optional overseas tax identifier values to the writes method.

Reply:
done